### PR TITLE
fbc-tests Test Suite update Mar 2018

### PR DESCRIPTION
**Pull request by jayrm**
- restructure test suite to use SUITE, TEST_GROUP, TEST framework from fbcunit
- fbc-tests output should be consistent across all unit-tests i.e. tests.fbc_tests.directory.file.suite.group.test
- fbcunit library updated to version 0.7
- some issues that existed since before this update marked in the source with TODO tags (approx 10)
- added ENABLE_CHECK_BUGS=1 makefile option to compile in known bugs, i.e. tests known to fail

### CODE REVIEW

**Comment by jayrm** [comment]
From the last commit, can see that the init proc is not getting executed.  At a loss to explain why.  Have tested it on dos32, win32, win64, linux64, and all works for me.  Not sure what is different in the Travis-CI build.

**Comment by dkl** [comment]
Hi Jeff, impressive work converting/cleaning up all the tests.

I tested here on linux-x86_64 and it crashed too. I ran `./fbc-tests --verbose` and noticed that it apparently crashes while running `fbc_tests.pointers.arith`:
```
  fbc_tests.pointers.arith
    ptrToFuncArray
Segmentation fault (core dumped)
```
However `ptrToFuncArray` is from `fbc_tests.pointers.array_ptr_fn`. So it looks like the suites and tests got mixed up.

**Comment by jayrm** [comment]
Thanks, dkl.  All the changes were manually made.  I really did try to read all the tests.  I like the generators; math-torture is excellent.  I was going through the directories alphabetically; when I got to ./structs/ I thought it would never end.

Thanks for testing.  Fixed now.  Your response about the suites and tests getting mixed up pointed me to the problem immediately.  I actually thought about this problem in the early stages of fbcunit library, but the fbc tests suites weren't rewritten yet.  Too bad I completely forgot about it!  Obviously different build systems handle module ctors/dtors differently.  :/

I am open to suggestions; with the number of changes, probably looking to see if the overall patterns make sense would be reasonable.

There are some TODO's in both fbcunit library and fbc-tests suite which I have documented or tagged.  I could keep adding to this pull request or merge this one and continue the changes in the next update.

**Comment by jayrm** [comment]
Just for fun, after much git diff & apply magic, I have pushed an alternate history of this patch to https://github.com/jayrm/fbc/tree/fbc-tests-alt .  The fbc-tests and fbc-tests-alt branches have the same end result, the difference in fbc-tests-alt is that all the white spaces changes are in the last commit.  The white space changes on their own are 50K+ lines.  The overall style & formatting of the code could still be improved, but I'm OK with it for now.

Thanks for the review notes.  They look straightforward to fix for this pull request and I will reply before I merge.  I like the idea about deriving the suite name from file name.  I'd like to leave that feature for the next update as it will hit every unit-test file.

File: tests/dim/common-shared-1.bas [hunk 1]
```diff
@@ -59,8 +59,11 @@ namespace shadowingACommon
 	end sub
 end namespace
 
-private sub ctor( ) constructor
-	fbcu.add_suite( "fbc_tests.dim.common-shared-1" )
-	fbcu.add_test( "test", @test_proc )
-	fbcu.add_test( "shadowing a COMMON", @shadowingACommon.test_proc )
-end sub
+SUITE( fbc_tests.dim_ )
```
[the remark this answers is not in the document]
Comment (jayrm) [reply tests/dim/common-shared-1.bas]: fixed.  tests involving ```common shared``` and ```extern``` and ```namespace``` may span multiple modules.  Added comments to show connections between modules.

File: tests/fbcunit/inc/fbcunit.bi [hunk 2]
```diff
@@ -223,13 +255,13 @@
 		#endmacro
 
 		#macro END_SUITE_EMIT( suite_name, id )
-				private sub suite_name##_ctor##id cdecl () constructor
+				private sub suite##_ctor##id cdecl () constructor
```
[the remark this answers is not in the document]
Comment (jayrm) [reply tests/fbcunit/inc/fbcunit.bi]: fixed, what should have been ```suite_ctor##id```.

File: tests/namespace/redim-1.bas [hunk 3]
```diff
@@ -14,34 +14,28 @@ namespace foo
 	end sub
 end namespace
 
-namespace fbc_tests.ns.redim_
+SUITE( fbc_tests.namespace_.redim_ )
```
[the remark this answers is not in the document]
Comment (jayrm) [reply tests/namespace/redim-1.bas]: in the first round of changes, was very cautious about changing the names of ```namespace```s especially in the tests that specifically test ```namespace```s.  after review, yeah, ```redim-1.bas```, ```redim-2.bas```, ```redim-2.bas``` are independent from one another.

**Comment by jayrm** [comment]
The first 4 jobs in each Travis-CI builds #105 & #106, fails to download oracle-java8-installer & oracle-java9-installer package during the container setup, and then ultimately fails the build process when unable to setup the package(s).  All this is in the build logs and don't see anywhere we control this, so I'm going to wait a few hours, then push a commit to this pull request to trigger another build.  

If everything builds OK then, I will merge this pull request.

**Comment by jayrm** [comment]
Several hours later, same end result but with different causes (e.g. unable to connect to apache.org) and Tavis-CI build #107 fails for the 32-bit tests when trying to setup the test container.  The 64-bit builds pass.  Travis-CI is great; I love it.  But what to do when the testing system itself fails?  I have a suspicion that we are not the only project with this problem right now, and it might take a while for the network to settle down.  My opinion is to merge this pull request, but I am not only developer here.

**Comment by dkl** [comment]
Yea, I agree, we should just merge anyways.

**Pull request merged**